### PR TITLE
Cache dynamic dropdown options

### PR DESCRIPTION
**Pull request by samelhusseini**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

https://github.com/google/blockly/issues/2493
https://github.com/google/blockly/issues/2926

### Proposed Changes
Only generate options during inialization (as we need the first tuple) and when a use clicks on the field to open it. (ie: when the editor is shown / widget is created). 

### Reason for Changes

Be more efficient about calling a dynamic dropdown function and not break in places where we expect certain options to be available. 

### Test Coverage

Tested the playground, and added a new test dropdown block that returns a random number of items every time the generate method is called. 

Tested on:
* Desktop Chrome

### CODE REVIEW

**Comment by samelhusseini** [comment]
@BeksOmega FYI.

I'm also a little surprised that we don't support an empty options list, but that can be dealt with in a different change.

File: core/field_dropdown.js [hunk 1]
```diff
@@ -421,15 +430,18 @@ Blockly.FieldDropdown.prototype.isOptionListDynamic = function() {
 
 /**
  * Return a list of the options for this dropdown.
+ * @param {boolean=} opt_regenerate Whether or not to re-generate the options.
  * @return {!Array.<!Array>} Array of option tuples:
  *     (human-readable text or image, language-neutral name).
  * @throws If generated options are incorrectly structured.
  */
-Blockly.FieldDropdown.prototype.getOptions = function() {
+Blockly.FieldDropdown.prototype.getOptions = function(opt_regenerate) {
```
Comment (BeksOmega) [inline core/field_dropdown.js]: `opt_regenerate` probably needs to default to true for backwards compat.

File: core/field_dropdown.js [hunk 2]
```diff
@@ -617,6 +629,9 @@ Blockly.FieldDropdown.validateOptions_ = function(options) {
           tuple);
     }
   }
+  if (!options.length) {
```
Comment (BeksOmega) [inline core/field_dropdown.js]: Thank you! Uncaught errors are the worst.

File: core/field_dropdown.js [hunk 3]
```diff
@@ -68,15 +68,23 @@ Blockly.FieldDropdown = function(menuGenerator, opt_validator) {
   this.menuGenerator_ = menuGenerator;
 
   /**
-   * The currently selected index. A value of -1 indicates no option
-   * has been selected.
+   * A cache of the most recently generated options.
+   * @type {Array.<!Array.<string>>}
+   * @private
+   */
+  this.generatedOptions_ = null;
```
Comment (BeksOmega) [inline core/field_dropdown.js]: Did you look into having this be a map/dictionary? I think there are some loops that could be elliminated that way.

File: core/field_dropdown.js [hunk 4]
```diff
@@ -68,15 +68,23 @@ Blockly.FieldDropdown = function(menuGenerator, opt_validator) {
   this.menuGenerator_ = menuGenerator;
 
   /**
-   * The currently selected index. A value of -1 indicates no option
-   * has been selected.
+   * A cache of the most recently generated options.
+   * @type {Array.<!Array.<string>>}
```
Comment (BeksOmega) [inline core/field_dropdown.js]: I think `.` in array declarations is deprecated.

File: tests/blocks/test_blocks.js [hunk 5]
```diff
@@ -1210,3 +1210,21 @@ Blockly.TestBlocks.removeDynamicDropdownOption_ = function() {
     }
   })
 };
+
+Blockly.Blocks['test_dropdowns_dynamic_random'] = {
```
Comment (BeksOmega) [inline tests/blocks/test_blocks.js]: Looks fun hehe.

**Comment by BeksOmega** [comment]
> I'm also a little surprised that we don't support an empty options list, but that can be dealt with in a different change.

Do you have a use case in mind? I think that's it generally good to oppose that behavior so that you avoid dealing with nulls in code generation. I'm having trouble thinking of a situation where the limitation can't be designed around (e.g. variables), but I could be missing something.

**Comment by RoboErikG** [comment]
> Do you have a use case in mind? I think that's it generally good to oppose that behavior so that you avoid dealing with nulls in code generation. I'm having trouble thinking of a situation where the limitation can't be designed around (e.g. variables), but I could be missing something.

I mentioned to @samelhusseini that I think our APIs made this easy to do with dynamic dropdown functions. Looking at our existing code, an empty options list is allowed by validateOptions_ but then we assume it's non-empty a few lines later in the constructor. It's probably better if we handle an empty list in one spot instead of expecting every option generator to check for it.

File: core/field_dropdown.js (hunk 4 above)
Comment (samelhusseini) [reply core/field_dropdown.js]: It is, but we're keeping it in Blockly for consistency.

**Comment by BeksOmega** [comment]
>  It's probably better if we handle an empty list in one spot instead of expecting every option generator to check for it.

I agree that it is a bit annoying to have to design around this limitation (as with most limitations hehe) but I think allowing empty lists and nulls would actually end up creating /more/ work (for us, and outside developers) in the longrun.

Firstly, if outside developers don't handle this at option-generation-time they'll end up having to handle it at code-generation-time. That is a much trickier spot to find bugs. I think we're actually helping them by notifying them before it gets to that point.

Secondly, allowing empty lists and nulls creates a lot more work for us, because now we have to check for that.

We'd have to start checking for empty lists all over the dropdown field, and special case them. For example:
  * Constructor would have to pass something different to the super.
  * getText_ would have to return `''` or `BLANK` or some placeholder.
  * And render_ would have to special-case somehow.
  
We'd also no longer be able to confidently call `field.getValue()` in a generic context, because it could return null. Making getValue() non-null saves us a lot of work, because it means you can remove `if(field.getValue())` checks.

Two other small points:
1) That `getOptions()[0]` line has been around since the [dawn of Blockly](https://github.com/google/blockly/blame/89af53a893185b269d9db635ce5cde29237cf5f6/core/field_dropdown.js#L46), and people seem mostly cool with it.
2) Displaying an empty dropdown is pretty buggy looking behavior anyway. If I were using a block programming language, I'd report that.

In Conclusion:
I think that throwing errors for empty lists is the proper way to handle things.

It saves outside developers the headache in the long run. It makes Blockly less fragile and more elegant. And I can't think of a usecase for empty lists (at least one that doesn't have a better solution).

**Comment by RoboErikG** [comment]
@BeksOmega good points, especially about it never working with an empty list. If @samelhusseini agrees then I'm okay with requiring a non-empty list of options and updating validateOptions_ to explicitly check for an empty list and throw a meaningful error in that case.

**Comment by samelhusseini** [comment]
I agree with the points made here. This PR adds a check in ``validateOptions`` and throws if the list returned is empty. I'll add further code comments so that we're more explicit about this.

File: core/field_dropdown.js (hunk 3 above)
Comment (samelhusseini) [reply core/field_dropdown.js]: What would the key of this map be? If it's the value, it means you can't have multiple options with the same value, label, is the same. I don't think we set that restrictions right now.

File: core/field_dropdown.js (hunk 1 above)
Comment (samelhusseini) [reply core/field_dropdown.js]: Yeah that makes sense.

File: core/field_dropdown.js [hunk 6]
```diff
@@ -421,15 +429,19 @@ Blockly.FieldDropdown.prototype.isOptionListDynamic = function() {
 
 /**
  * Return a list of the options for this dropdown.
- * @return {!Array.<!Array>} Array of option tuples:
+ * @param {boolean=} opt_useCache For dynamic options, whether or not to use the
+ *     cached options or to re-generate them.
+ * @return {!Array.<!Array>} A non-empty array of option tuples:
  *     (human-readable text or image, language-neutral name).
  * @throws If generated options are incorrectly structured.
```
Comment (alschmiedt) [inline core/field_dropdown.js]: nit: Could we add the type here? @throws {Type}

Comment (samelhusseini) [reply core/field_dropdown.js]: Done.

File: core/field_dropdown.js (hunk 3 above)
Comment (BeksOmega) [reply core/field_dropdown.js]: > What would the key of this map be?

It would be keyed by value. This is what the dropdown currently uses, it's just inefficient and uses for-loops instead of a map.

We may not currently throw errors, but the language-neutral keys of your dropdown have to be unique, otherwise your dropdown breaks:

SameValues

<details><summary>Steps to reproduce</summary>
<p>

1. Add the following block definition to the playground.
```
{
    "type": "test_dropdowns_same_value",
    "message0": "same value %1",
    "args0": [
      {
        "type": "field_dropdown",
        "name": "FIELDNAME",
        "options": [
          [ "first item", "ITEM1" ],
          [ "second item", "ITEM2" ],
          [ "third item", "ITEM2" ],
          [ "fourth item", "ITEM2" ],
          [ "fifth item", "ITEM2" ],
        ]
      }
    ],
    "style": "fields_dropdowns"
  }
```
2. Add the block to the workspace.
3. Use the dropdown.
4. Observe the bad behavior.
</p>
</details>

If we were to switch it to a map we could definitely get rid of the loop in `doClassValidation_`. We could also probably get rid of the `selectedIndex_` property and just use the value instead, elliminating the loop in `doValueUpdate_`.
